### PR TITLE
[codex] Add KiwiSDR protocol metadata scaffolding

### PR DESCRIPTION
**Pull request by rfoust**
## Summary

Adds KiwiSDR protocol/capability scaffolding, richer read-only receiver metadata, diagnostic compressed SND/W/F support, automation snapshots, and monitor/camping state handling. Full-but-API-capable receivers now continue to WebSocket admission so `MSG monitor` can offer waiting/camping, `qpos` channel-free notifications are actionable via Connect, and monitor/camping sessions no longer present a controllable Kiwi waterfall.

The automation bridge now exposes `get kiwi` / `get kiwisdr`, returning Kiwi profile state, metadata, protocol capability state, stream observations, sequence counters, waterfall availability, and diagnostic compression request state for future non-pixel automation checks.

## Constitution principle honored

Principle IV — clean-room contributions. KiwiSDR source was used only as a protocol reference where license-compatible, and provenance is recorded in `docs/kiwisdr-cleanroom-design.md`.

Consulted KiwiSDR reference:
- Repository: `https://github.com/jks-prv/KiwiSDR.git`
- Commit: `a83085fe2222dd3e374910faf2195e0454b556ae`
- Files: `rx/rx_waterfall.h`, `rx/rx_waterfall.cpp`, `rx/rx_cmd.h`, `rx/rx_cmd.cpp`, `rx/rx_sound.h`, `rx/rx_sound.cpp`, `rx/rx_sound_cmd.cpp`, `rx/csdr/ima_adpcm.h`, `rx/csdr/ima_adpcm.cpp`, `rx/rx_server.cpp`, `rx/rx_monitor.cpp`, `rx/rx_server_ajax.cpp`
- `web/kiwi/monitor.js` was viewed only to confirm user-facing monitor/camp wording; no JavaScript code was copied or translated.

## Test plan

- [x] Local build passes (`cmake --build build --target AetherSDR kiwi_sdr_protocol_test -j8`)
- [x] Behavior verified on a real radio if applicable: N/A, KiwiSDR receive-only path; no Flex radio-side commands are sent for Kiwi profile state
- [ ] Existing tests pass (CI)
- [x] Reproduction steps documented if user-reported bug: covered by non-network protocol regressions and automation bridge smoke

Additional local validation:
- [x] `ctest --test-dir build -R 'kiwi_sdr_protocol_test|kiwi' --output-on-failure`
- [x] `git diff --check`
- [x] `python3 tools/check_a11y.py` exited 0; only existing warning-only findings were reported
- [x] Automation bridge smoke with `QT_QPA_PLATFORM=offscreen AETHER_AUTOMATION=1 ./build/AetherSDR.app/Contents/MacOS/AetherSDR`, then `python3 tools/automation_probe.py ping` and `python3 tools/automation_probe.py get kiwi`

## Checklist

- [x] Commits are signed (`docs/COMMIT-SIGNING.md`)
- [x] No new flat-key `AppSettings` calls — use nested-JSON-under-one-key (Principle V)
- [x] Code is clean-room — not decompiled, disassembled, or reverse-engineered from a proprietary binary (Principle IV)
- [x] All meter UI uses `MeterSmoother` (AGENTS.md convention)
- [x] Documentation updated if user-visible behavior changed, including `docs/automation-bridge.md`
- [x] Security-sensitive changes reference a GHSA if applicable: N/A

### CODE REVIEW

File: src/gui/RadioSetupDialog.cpp [hunk 1]
```diff
@@ -3459,6 +3600,13 @@ QWidget* RadioSetupDialog::buildAntennaNamesTab()
                 (*refreshKiwi)();
             }
         });
+        connect(m_kiwiSdrManager, &KiwiSdrManager::profileTelemetryChanged,
+                this, [refreshKiwi](const QString&,
+                                    const KiwiSdrReceiverTelemetry&) {
+            if (*refreshKiwi) {
+                (*refreshKiwi)();
+            }
+        });
```
[the remark this answers is not in the document]
Comment (rfoust) [reply src/gui/RadioSetupDialog.cpp:3503]: Addressed in 6d14ccac: telemetry-driven `profileTelemetryChanged` updates now arm a single-shot 150 ms timer so bursts coalesce into one Kiwi row rebuild. Profile list and connection-state changes still refresh immediately.

File: src/gui/MainWindow_KiwiSdr.cpp [hunk 2]
```diff
@@ -384,6 +745,54 @@ QJsonObject MainWindow::automationSetSliceReceiveSource(const QString& arg)
...
+            });
+        }
+    }
+
+    return QJsonObject{
+        {QStringLiteral("diagnosticWaterfallCompressionRequested"),
+            KiwiSdrClient::diagnosticWaterfallCompressionRequested()},
+        {QStringLiteral("profiles"), profiles},
+        {QStringLiteral("profileCount"), profiles.size()},
+    };
```
[the remark this answers is not in the document]
Comment (rfoust) [reply src/gui/MainWindow_KiwiSdr.cpp:795]: Addressed in 6d14ccac: `automationKiwiSdrSnapshot()` now exposes `diagnosticSoundCompressionRequested` alongside the waterfall flag, and `docs/automation-bridge.md` documents both top-level diagnostic fields.

**Comment by rfoust** [comment]
Follow-up update pushed in `bd4eef2d`:

- Waiting/camping text now says AetherSDR is waiting for a free KiwiSDR receiver slot and will reconnect automatically when the server reports one is available.
- `KiwiSdrManager` now schedules the existing 3-second normal reconnect path when `MSG qpos` metadata reports the reload/free-slot hint, including manual Connect attempts that are sitting in `Waiting`.
- The Kiwi setup row keeps `Waiting` sessions as active, so the button remains `Disconnect` to cancel the wait/retry instead of switching back to `Connect`.
- The clean-room design note now records this queue boundary.

Validation:
- `cmake --build build --target AetherSDR kiwi_sdr_protocol_test -j8`
- `ctest --test-dir build -R 'kiwi_sdr_protocol_test|kiwi' --output-on-failure`
- `git diff --check`
- `python3 tools/check_a11y.py` (warning-only existing findings)
- `QT_QPA_PLATFORM=offscreen AETHER_AUTOMATION=1 ./build/AetherSDR.app/Contents/MacOS/AetherSDR`
- `python3 tools/automation_probe.py ping`
- `python3 tools/automation_probe.py get kiwi`

**Comment by rfoust** [comment]
Follow-up update pushed in `f91b903a`:

- The waterfall overlay now uses queue-aware wording during Kiwi monitor/waiting sessions instead of the old generic “monitor sessions do not provide a controllable waterfall” message.
- Waiting overlays now title as “Waiting for KiwiSDR receiver slot” and explain that AetherSDR will reconnect automatically when the server reports a slot is available.
- Queue position, free-slot/reconnecting, and camped-monitor cases each get specific waterfall detail text.

Validation:
- `cmake --build build --target AetherSDR kiwi_sdr_protocol_test -j8`
- `ctest --test-dir build -R 'kiwi_sdr_protocol_test|kiwi' --output-on-failure`
- `git diff --check`
- `python3 tools/check_a11y.py` (warning-only existing findings)
- `QT_QPA_PLATFORM=offscreen AETHER_AUTOMATION=1 ./build/AetherSDR.app/Contents/MacOS/AetherSDR`
- `python3 tools/automation_probe.py ping`
- `python3 tools/automation_probe.py get kiwi`

File: src/core/KiwiSdrProtocol.cpp [hunk 3]
```diff
@@ -172,6 +498,819 @@ QVector<MsgToken> parseMsgTokens(const QString& message)
...
+        const int value = parseInt(&ok);
+        if (ok) {
+            changed = setIntField(&metadata->extApi, &metadata->hasExtApi,
+                                  value) || changed;
+        }
+    } else if (key == QStringLiteral("too_busy")) {
+        structured = true;
+        bool ok = false;
+        const int value = parseInt(&ok);
+        const bool busy = !ok || value != 0;
```
Comment (ten9876) [inline src/core/KiwiSdrProtocol.cpp]: **[blocker]** `const bool busy = !ok || value != 0;` forces `busy=true` whenever the integer fails to parse, so a malformed/empty `too_busy=` (or an injected MSG line) marks a reachable receiver as busy and connect/queue logic may abandon it. Set busy only on a successful parse:
```cpp
const int value = parseInt(&ok);
if (ok)
    changed = setBoolField(&metadata->busy, &metadata->hasBusy, value != 0) || changed;
```

File: src/gui/MainWindow_KiwiSdr.cpp [hunk 4]
```diff
@@ -1195,7 +1620,7 @@ void MainWindow::wireKiwiSdr()
                 this, [this](const QString& profileId, KiwiSdrClient::State state,
                              const QString&) {
             QString panId;
```
Comment (ten9876) [inline src/gui/MainWindow_KiwiSdr.cpp:1624]: **[blocker — read-only safety]** This branch now fires for `Camping`, so `setTrackedSlice`/`setReceiverControls` are pushed into the client during a monitor/camp session. The *only* thing stopping tune/mode commands from reaching the read-only camped receiver is the client-side `m_monitorMode` early-return in `sendTrackedSliceToServer`/`sendReceiverControlsToServer`. That's a single load-bearing invariant with nothing tying `m_monitorMode` to the manager's `Camping` state. Please gate this push on an explicit `controllable(state)` predicate (or assert `m_monitorMode` whenever state is `Camping`/`Waiting`) so a future camp-accept/reconnect path can't desync and leak control.

File: src/gui/MainWindow_Wiring.cpp [hunk 5]
```diff
@@ -1642,23 +1620,19 @@ void MainWindow::wirePanadapter(PanadapterApplet* applet)
...
                     slice->sliceId(), applet->panId(), centerMhz,
                     bandwidthMhz, sw->wfLineDuration());
-                if (profileId == displayProfileId
-                    && m_kiwiSdrManager->isConnected(profileId)) {
-                    sw->setKiwiSdrWaterfallAvailable(true);
+                const bool profileCanDriveWaterfall =
+                    KiwiSdrClient::stateAllowsReceiverControl(
+                        m_kiwiSdrManager->state(profileId))
+                    && m_kiwiSdrManager->waterfallAvailable(profileId);
+                if (profileId == displayProfileId && profileCanDriveWaterfall) {
```
Comment (ten9876) [inline src/gui/MainWindow_Wiring.cpp:1625]: **[blocker — consistency]** `profileCanDriveWaterfall` gates on `stateAllowsReceiverControl` (Connected-only), but `isConnected()`/`stateHasReceiveAudio()` elsewhere treat `Camping` as connected. So a camped profile is "connected" for some decisions (e.g. `assignSliceToProfile`) yet can never drive/refresh its waterfall here, and zoom/pan stops forwarding geometry → the aperture desyncs. Centralize this in one per-state capability table (controllable / has-audio / has-waterfall / label) consumed by AutomationServer, KiwiSdrManager and the GUI, instead of re-deriving "connected" three different ways.

File: src/core/KiwiSdrClient.cpp [hunk 6]
```diff
@@ -654,15 +660,11 @@ void KiwiSdrClient::handleStatusPreflightFinished(QNetworkReply* reply)
             << "users_max="
```
Comment (ten9876) [inline src/core/KiwiSdrClient.cpp:662]: **[blocker — UX]** The full-receiver early-abort (`setState(Error)`+`cleanupSockets()` with the precise `(N/M users) at capacity` message) was removed so full-but-API-capable receivers can proceed to admission. But a receiver that is full **and not** camp/monitor-capable now also proceeds, sits in `Connecting`, and only fails when the audio-ready timeout fires — a slower, vaguer failure that drops the capacity message. Keep the immediate abort for the full-and-not-camp-capable case.

File: src/core/KiwiSdrClient.cpp [hunk 7]
```diff
@@ -1739,9 +1760,34 @@ void KiwiSdrClient::handleSoundFrame(const QByteArray& frame)
         meterReading.squelched = header.squelched;
     }
     markSoundAudioReady();
```
Comment (ten9876) [inline src/core/KiwiSdrClient.cpp:1764]: **[fast-follow — `AETHER_KIWI_SND_COMP` only]** `markSoundAudioReady()` flips the profile to Connected/Camping *before* the compressed-SND path below can early-return on an unsynchronized ADPCM frame (sequence gap, no restart bit). If the server never sends a restart-flagged frame the session reports Connected/Monitoring but emits zero PCM indefinitely. Defer the receive-ready transition until the first frame that actually yields samples, at least on the compressed path.

File: src/core/KiwiSdrProtocol.cpp [hunk 8]
```diff
@@ -172,6 +498,819 @@ QVector<MsgToken> parseMsgTokens(const QString& message)
...
+}
+
+void invalidateSoundAdpcmState(SoundAdpcmState* state)
+{
+    if (state) {
+        state->valid = false;
+    }
+}
+
+SoundFrameDecodeResult decodeSoundFrame(const QByteArray& frame,
```
Comment (ten9876) [inline src/core/KiwiSdrProtocol.cpp:729]: **[fast-follow — `AETHER_KIWI_SND_COMP` only]** `SoundAdpcmState` is default-constructed `valid{true}` and `handleSoundFrame` only drops when `!valid`, so the *first* compressed frame with the restart bit clear decodes IMA-ADPCM from `predictor=0/index=0` rather than being dropped — a burst of corrupted audio until a restart frame resyncs. Start the state `valid=false` (decode only after the first restart), or gate the first-frame decode on the restart bit. (Related: the pre-existing `sequence = counter & 0xffu` in `parseSoundFrameHeader` makes 256-multiple gaps invisible to the resync this newly relies on.)

**Review by ten9876 — DISMISSED** [review]
## Review: changes requested

High-effort pass over the Kiwi protocol scaffolding (8 finder angles + adversarial verification). CI is green, conventions are clean (receive-only path, no flat-key settings, a11y names present), and the binary parser is genuinely defensive — no crash/UAF/overflow found. Findings are inline; scope notes below.

### Blockers (inline)
- `too_busy` parse-failure forces `busy=true` — default MSG path.
- The camped-receiver read-only safety rests entirely on the client `m_monitorMode` guard, with nothing binding it to the manager `Camping` state.
- Two definitions of "connected" (control-capable vs has-receive-audio) drive inconsistent waterfall/aperture decisions for camped profiles.
- The full-receiver fail-fast was dropped for the full-AND-not-camp-capable case.

### Fast-follow (inline, `AETHER_KIWI_SND_COMP`-gated diagnostic path)
Connected-but-silent on a no-restart compressed frame, and predictor-0 first-frame ADPCM corruption. Low reachability (diagnostic env var), but real.

### Pre-existing — NOT introduced by this PR (recommend a separate issue, not blocking)
- `extractMeterFromSndVerifiedLayout` (main line 263) emits a **Verified**-confidence S-meter from any 1034-byte SND frame without checking the compression flag; under compression it reads ADPCM init bytes as a calibrated meter. The fix is to consult the `MeterContext.compressionObserved` it already receives but `Q_UNUSED`s.
- `parseSoundFrameHeader` collapses the 32-bit sequence counter to its low byte (main line 68) — this PR's new ADPCM resync now depends on it, so it's worth widening alongside the compressed work.

### Cleanup / altitude (non-blocking)
- Per-frame double/triple parse on the hot path: each SND header is decoded ~4× and each W/F frame inspected ~3× (build the header/shape once and pass it down).
- Duplicate IMA-ADPCM nibble decoders; hand-rolled endian readers where `qFromLittleEndian`/`qFromBigEndian` already exist; a dead pass-through `diagnosticWaterfallCompressionFlagEnabled`.
- Speculative scaffolding the PR itself labels as such: the 8-state `CampStatus` enum with no camp *connect* path, and the `Q_UNUSED` `MeterContext`.
- A latent two-parsers-for-one-header hazard: `parseWaterfallFrameHeader` reads byte[8], `inspectWaterfallFrame` reads a u16+compression flag at offset 8 — they agree for zoom≤30 but diverge on a malformed frame. Worth collapsing to one parser.

Happy to pair on the `m_monitorMode` / state-capability-table consolidation — that's the one with on-air-adjacent risk. Thanks for the clean-room provenance and the automation-bridge test notes.

**Comment by ten9876** [comment]
### Follow-up review pass (HEAD `8c804f0a`)

Re-ran the high-effort review against HEAD. The four original blockers are resolved and verified, plus two more parse-safety items were folded in:

- `c2ce492a` — original blockers (too_busy default, camped read-only safety via `receiverControlSuppressed()`, capacity message)
- `127fe297` — reset `m_preflightReportedFull` on the wss retry (it was only reset in `connectToEndpoint()`, so a ws→wss retry timeout could emit a stale capacity message)
- `8c804f0a` — `qpos` now validates position + waiters before applying the queue update (+ regression test), matching the `too_busy` discipline

Clean-room provenance and the receive-only guarantee (Principle VI) both check out. CI is green.

### Outstanding follow-ups (non-blocking — leaving these to you, @rfoust)

**Correctness (low reachability):**
1. **Waterfall 517-byte misclassify** — `classifyWaterfallFrame` (`KiwiSdrProtocol.cpp` ~L827) treats a 517-byte zoomed row as ADPCM-compact via `payloadBytes==517` even when the parsed compression flag is false; a direct/uncompressed row of that exact shape would be ADPCM-decoded into garbage. The flag is already parsed in `shape.compressed` — trust it instead of OR-ing the size heuristic.
2. **`metadataValueToMhz` sub-1 MHz scaling** (~L230) — the magnitude heuristic divides `[1000,1e6)` by 1000 as "kHz", so an LF/MW Kiwi reporting a sub-1 MHz `center_freq` displays 1000× high. Standard 0–30 MHz receivers are unaffected. Convert from the known unit rather than guessing by magnitude.

**Architecture (the on-air-adjacent one):**
3. **Camp/monitor state lives in parallel flags** (`m_monitorMode` / `m_campAccepted` / `m_monitorQueueRequested` / `m_preflightReportedFull`) rather than a single source of truth driven from `metadata.campStatus`. `receiverControlSuppressed() = m_monitorMode || state==Camping` is itself defensive against those flags disagreeing, and the frame-vs-`camp`-MSG ordering (briefly lands in `Connected` before self-correcting to `Camping`) is another symptom. Happy to pair on consolidating this into a single camp-session state — it's the one with read-only-safety implications.

**Cleanup (minor):** per-frame header re-parse on the SND/WF decode path (classify+decode both parse), a couple of O(n) per-frame scans (`observedLayouts.contains`, unsupported-frame dedup), the `Q_UNUSED` `MeterContext`, and hand-rolled endian readers where `qFromLittleEndian`/`qFromBigEndian` exist.

None of these block merge.

**Review by ten9876 — APPROVED** [review]
Approved. High-effort review across three passes (8 finder angles + adversarial verification each): the binary/protocol parser is genuinely defensive (no crash/UAF/overflow), clean-room provenance is solid, and the receive-only/Principle VI guarantee holds via receiverControlSuppressed(). Blockers were fixed and verified (too_busy default, camped read-only safety, capacity message, preflight-flag reset, qpos parse-safety). Remaining items — waterfall-517 ADPCM misclassify, sub-1MHz metadataValueToMhz, and the camp-flag→state-machine consolidation — are non-blocking and documented on the PR as follow-ups for the author. CI green (build/macOS/Windows).
